### ISSUE TITLE
Why does NSPredicate accessibilityIdentifier fail?

### ISSUE THREAD

**Issue by bjcullinan** [bricepollock/swift-snippets#1]
I asked this question on SO with no response so I thought I would post it here too since it mentions your article.  Your article is one of the few pages I could find on Google that actually provides examples on this.
When I run a test with the following code:

```
let shuffle = app.buttons.elementMatchingPredicate(NSPredicate(format: "accessibilityIdentifier=%@", "shuffle"))
expectationForPredicate(NSPredicate(format: "hittable=TRUE"), evaluatedWithObject: shuffle, handler: nil)
waitForExpectationsWithTimeout(2) {_ in}
shuffle.tap()

```
The test fails. But when I use this code:

```
let shuffle = app.buttons["shuffle"]
expectationForPredicate(NSPredicate(format: "hittable=TRUE"), evaluatedWithObject: shuffle, handler: nil)
waitForExpectationsWithTimeout(2) {_ in}
shuffle.tap()
```

http://www.bricepollock.com/difference-xcode-ui-testing-ui-automation-instrument/

This guy says specifically that a huge improvement to UIT over UIA is that UIT uses queries that are quickly evaluated every time it is used, instead of evaluating once and operating on the result set. It seems there is a poor design principle here. Can anyone explain what is wrong with the first query?

`expectationForPredicate(NSPredicate("ANY accessibilityIdentifier=%@"), evaluatedWithObject: app.buttons)`
Doesn't seem to work either. Is there documentation that describes which functions are queries and which are result sets? I could only find 1 page on Apple's site that describes how to use this new XCUI suite, and it is severely limited on examples.

**Comment by bpollock-coursera** [comment]
I'd recommend watching the WWDC video:
https://developer.apple.com/videos/play/wwdc2015-406/

...and reading a bunch of articles on the subject:
https://www.bignerdranch.com/blog/ui-testing-in-xcode-7-part-1-ui-testing-gotchas/

This place looks to have the documentation on the XCUI... objects you'll use in unit testing:
http://masilotti.com/xctest-documentation/

I don't have a specific answer for why something is failing because that lies in part on your UI. Looks like you are setting the accessibility identifier which looks correct. I would write this test like so:

let shuffle = app.buttons["shuffle"]
XCTAssert(shuffle.hittable == true)
shuffle.tap()

Timeouts aren't typically necessary anymore with the Xcode UI Testing framework.

**Comment by bricepollock** [comment]
I'd recommend watching the WWDC video:
https://developer.apple.com/videos/play/wwdc2015-406/

...and reading a bunch of articles on the subject:
https://www.bignerdranch.com/blog/ui-testing-in-xcode-7-part-1-ui-testing-gotchas/

This place looks to have the documentation on the XCUI... objects you'll use in unit testing:
http://masilotti.com/xctest-documentation/

I don't have a specific answer for why something is failing because that lies in part on your UI. Looks like you are setting the accessibility identifier which looks correct. I would write this test like so:

let shuffle = app.buttons["shuffle"]
XCTAssert(shuffle.hittable == true)
shuffle.tap()

Timeouts aren't typically necessary anymore with the Xcode UI Testing framework.

**Comment by bjcullinan** [comment]
Great advice, thank you guys.

**Comment by bjcullinan** [comment]
Here is the SO question if you want to copy and paste your response I will mark it correct and you can get the points: http://stackoverflow.com/questions/35283070/why-does-nspredicate-accessibilityidentifier-fail
